src/Utilities/lazy_iterators.jl

```diff
@@ -0,0 +1,65 @@
+struct EmptyVector{T} <: AbstractVector{T} end
+Base.size(::EmptyVector) = (0,)
+Base.isempty(::EmptyVector) = true
+Base.eltype(::EmptyVector{T}) where {T} = T
+Base.iterate(::EmptyVector) = nothing
+
+struct LazyMap{T, VT}
+    f::Function
+    data::VT
+end
+function LazyMap{T}(f::Function, data) where {T}
+    return LazyMap{T, typeof(data)}(f, data)
+end
+Base.size(it::LazyMap) = size(it.data)
+function Base.iterate(it::LazyMap, args...)
+    elem_state = iterate(it.data, args...)
+    if elem_state === nothing
+        return nothing
+    else
+        return it.f(elem_state[1]), elem_state[2]
+    end
+end
+Base.IteratorSize(it::LazyMap) = Base.IteratorSize(it.data)
+Base.eltype(::LazyMap{T}) where {T} = T
+
+struct LazyFilter{T, VT}
```

Comment:
Is this the same as `Base.Iterators.Filter`?

Reply:
Indeed, `filter` was not lazy but I didn't see `Base.Iterators.Filter`